### PR TITLE
Decouple resource acls from external group

### PR DESCRIPTION
**Pull request by mzmssg**
#3275

### CODE REVIEW

File: src/rest-server/src/config/v2/group.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ const groupCreateInputSchema = Joi.object().keys({
     .empty(''),
   extension: Joi.object()
     .pattern(/\w+/, Joi.required())
-    .default({}),
```
Comment (ydye) [inline src/rest-server/src/config/v2/group.js]: Please keep it

File: src/rest-server/src/config/v2/user.js [hunk 2]
```diff
@@ -74,7 +74,7 @@ const userCreateInputSchema = Joi.object().keys({
     .required(),
   extension: Joi.object()
     .pattern(/\w+/, Joi.required())
-    .default({}),
```
Comment (ydye) [inline src/rest-server/src/config/v2/user.js]: Keep it

Comment (mzmssg) [reply src/rest-server/src/config/v2/user.js]: For object, it's better to keep a empty default, see https://github.com/hapijs/joi/issues/1135

Comment (ydye) [reply src/rest-server/src/config/v2/user.js]: Can you test this API and request it without the extension field. Could share me with the result?

File: src/rest-server/src/utils/manager/user/user.js [hunk 3]
```diff
@@ -33,13 +33,13 @@ const userSchema = Joi.object().keys({
   password: Joi.string()
     .empty('')
     .default(''),
-  extension: Joi.object().pattern(/\w+/, Joi.required()).required(),
+  extension: Joi.object().pattern(/\w+/, Joi.required()).default(),
```
Comment (ydye) [inline src/rest-server/src/utils/manager/user/user.js]: This filed must be a ```{}```, if you think replace ```required``` with default is better. Please set default value as ```{}``` (empty dict). This schema constraint the final structure in secret, pls check the data strictly here. If you wanna set any default value logic, you'd better put them in the controller or model level.

File: src/rest-server/src/utils/manager/group/group.js [hunk 4]
```diff
@@ -29,13 +29,23 @@ const groupSchema = Joi.object().keys({
     .default(''),
   extension: Joi.object()
     .pattern(/\w+/, Joi.required())
-    .required(),
+    .keys({
+      acls: Joi.object()
```
Comment (ydye) [inline src/rest-server/src/utils/manager/group/group.js]: From the perspective of OpenPAI, I think this change is fine, because we could strictly ask each opertion add the acl filed to every group, or the operation will fail.

Comment (ydye) [reply src/rest-server/src/utils/manager/group/group.js]: But, I wonder if it is possible to remove this check into model level? Or in the next time, if we add new value into extension, we will have to modify the schema and add new entry into it again. Or do you have a better solution?

File: src/rest-server/src/utils/common.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+const assignValueByKeyarray = (obj, keyarray, value) => {
```
Comment (ydye) [inline src/rest-server/src/utils/common.js]: Pls add jsdoc here.

File: src/rest-server/src/routes/v2/group.js [hunk 6]
```diff
@@ -36,6 +36,10 @@ router.route('/:groupname/extension')
 /** Put /api/v2/group/:groupname/extension */
   .put(token.check, param.validate(groupInputSchema.groupExtensionUpdateInputSchema), groupController.updateGroupExtension);
 
+router.route('/:groupname/extension/*')
+/** Put /api/v2/group/:groupname/extension/:attribute */
+  .put(token.check, groupController.updateGroupExtensionAttr);
```
Comment (ydye) [inline src/rest-server/src/routes/v2/group.js]: param.validate

File: src/rest-server/src/models/v2/user.js [hunk 7]
```diff
@@ -86,35 +65,19 @@ const createUserIfNonExistent = async (username, userValue) => {
...
-      // admin has the permission of all groups.
-      ret = true;
-    }
-    return ret;
-  } catch (error) {
-    throw error;
-  }
+const getUserVCs = async (username) => {
+  const userItem = await getUser(username);
+  return groupModel.getGroupsVCs(userItem.grouplist);
```
Comment (ydye) [inline src/rest-server/src/models/v2/user.js]: The function named getUserVCs returns grouplist?

File: src/rest-server/src/models/v2/user.js [hunk 8]
```diff
@@ -19,59 +19,38 @@
...
-  }
+  return await crudUser.remove(username, crudConfig);
 };
 
+// it's an inplace encrypt!
 const getEncryptPassword = async (userValue) => {
-  try {
     await user.encryptUserPassword(userValue);
     return userValue;
-  } catch (error) {
```
Comment (ydye) [inline src/rest-server/src/models/v2/user.js]: This time is fine. But next time, I think the unrelative change should be in another PR.

File: src/rest-server/deploy/legacy_user_migrate.py [hunk 9]
```diff
@@ -138,6 +138,60 @@ def secret_data_prepare_v2_group(self, groupname):
         post_data_dict['data'] = group_dict
         return post_data_dict
 
+    def load_v2_groups(self):
+        ns_pai_group_list = self.list_all_secrets_from_namespace(self.secret_ns_group_v2)
+        decode_group_list = []
+        vc_set = set()
+        for group in ns_pai_group_list:
+            meta_dict = dict()
+            meta_dict['name'] = bytes.fromhex(group.metadata.name).decode('utf-8')
```
Comment (ydye) [inline src/rest-server/deploy/legacy_user_migrate.py]: Tag 1

File: src/rest-server/deploy/legacy_user_migrate.py [hunk 10]
```diff
@@ -138,6 +138,60 @@ def secret_data_prepare_v2_group(self, groupname):
...
+                extension_dict['acls']['virtualClusters'].extend(all_vcs)
+            elif extension_dict['groupType'] == 'vc':
+                extension_dict['acls']['virtualClusters'].append(data_dict['data']['groupname'])
+
+        return data_dict, True
+
+    def update_v2_group(self, data_dict):
+        post_data_dict = {
+            'metadata': {
+                'name': data_dict['metadata']['name'].encode('utf-8').hex()
```
Comment (ydye) [inline src/rest-server/deploy/legacy_user_migrate.py]: In tag1, I think it's unnecessary to do decode and then encode here. Why not directly use the value from k8s?

File: src/rest-server/deploy/legacy_user_migrate.py [hunk 11]
```diff
@@ -138,6 +138,60 @@ def secret_data_prepare_v2_group(self, groupname):
...
 
+    def load_v2_groups(self):
+        ns_pai_group_list = self.list_all_secrets_from_namespace(self.secret_ns_group_v2)
+        decode_group_list = []
+        vc_set = set()
+        for group in ns_pai_group_list:
+            meta_dict = dict()
+            meta_dict['name'] = bytes.fromhex(group.metadata.name).decode('utf-8')
+            group_dict = {
+                'groupname': str(base64.b64decode(group.data['groupname'].encode('utf-8')), 'utf-8'),
```
Comment (ydye) [inline src/rest-server/deploy/legacy_user_migrate.py]: Are there any update or change happening at groupname, description or externalName? If not, you can use the value from k8s directly. And I think decode is unnecessary.

File: src/rest-server/deploy/legacy_user_migrate.py [hunk 12]
```diff
@@ -295,32 +368,39 @@ def main():
...
+        vc_set = transferCli.vc_set
+        for vc in vc_set:
+            secret_post_data = transferCli.secret_data_prepare_v2_group(vc)
+            transferCli.create_secret_group_v2(secret_post_data)
+        logger.info('Legacy user data transfer from namespace v1 to namespace v2 successfully')
     else:
-      logger.info("Legacy data has already been transferred from v1 to v2. Skip it.")
+        logger.info("Legacy data has already been transferred from v1 to v2. Skip it.")
+
+    group_list, vc_set = transferCli.load_v2_groups()
```
Comment (ydye) [inline src/rest-server/deploy/legacy_user_migrate.py]: Will this logic happen everytime if we start rest-server? I have some concerns about the result if this script executed several times. Have you consider it?

File: src/rest-server/src/controllers/v2/group.js [hunk 13]
```diff
@@ -104,6 +104,29 @@ const updateGroupExtension = async (req, res, next) => {
   }
 };
 
+const updateGroupExtensionAttr = async (req, res, next) => {
+  try {
+    const groupname = req.params.groupname;
+    const attrs = req.params[0].split('/');
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/group.js]: params[0]？？？？？？？ Can you provide some schema about this request?

File: src/rest-server/src/controllers/v2/group.js [hunk 14]
```diff
@@ -104,6 +104,29 @@ const updateGroupExtension = async (req, res, next) => {
...
+    const updateData = req.body.data;
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/group.js]: The same problem. We have no idea about this data.

File: src/rest-server/src/controllers/v2/group.js [hunk 15]
```diff
@@ -168,4 +191,5 @@ module.exports = {
   updateGroupDescription,
   updateGroupExternalName,
   deleteGroup,
+  updateGroupExtensionAttr,
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/group.js]: So this function can only update one key field in extension, right?

File: src/rest-server/src/models/v2/group.js [hunk 16]
```diff
@@ -21,7 +21,6 @@ const authConfig = require('@pai/config/authn');
 const secretConfig = require('@pai/config/secret');
 const adapter = require('@pai/utils/manager/group/adapter/externalUtil');
 const config = require('@pai/config/index');
-const userModel = require('@pai/models/v2/user');
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: Is it necessary to do this change? Because you do the same operation in the function.

File: src/rest-server/src/models/v2/group.js [hunk 17]
```diff
@@ -31,236 +30,182 @@ const crudConfig = crudGroup.initConfig(process.env.K8S_APISERVER_URI);
...
 
-const createGroup = async (groupname, groupValue) => {
-  try {
-    const ret = await crudGroup.create(groupname, groupValue, crudConfig);
-    await addGroupIntoAdminUser(groupname, groupValue);
-    return ret;
-  } catch (error) {
-    throw error;
-  }
+const getGroupVCs = async (groupname) => {
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: getGroupsVCs(['groupname']) === getGroupVCs('groupname') ?

File: src/rest-server/src/models/v2/group.js [hunk 18]
```diff
@@ -31,236 +30,182 @@ const crudConfig = crudGroup.initConfig(process.env.K8S_APISERVER_URI);
...
+  return Promise.all(grouplist.map(getGroup));
+};
+
+const batchUpdateGroups = async (groupItems) => {
+  return await Promise.all(groupItems.map(async (groupItem) => {
+    await updateGroup(groupItem.groupname, groupItem);
+  }));
+};
+
+const getVCsWithGroupinfo = async (groupItems) => {
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: getVCsWithGroupInfo

File: src/rest-server/src/models/v2/group.js [hunk 19]
```diff
@@ -313,41 +258,31 @@ const initGrouplistInCfg = async () => {
...
-    logger.error('Failed to create admin group configured in configuration.');
+    logger.error('Failed to create group configured in configuration.');
     // eslint-disable-next-line no-console
     console.log(error);
   }
 };
 
-const createAdminUser = async () => {
+const createDefaultAdminUser = async () => {
+  const userModel = require('@pai/models/v2/user');
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: Requre in the top

**Comment by ydye** [comment]
- [ ] Deploy this code with basic mode, test all feature
- [ ] Deploy this code with AAD mode, test all feature

File: src/rest-server/src/controllers/v2/user.js [hunk 20]
```diff
@@ -20,13 +20,25 @@ const userModel = require('@pai/models/v2/user');
...
 
 const getUser = async (req, res, next) => {
   try {
     const username = req.params.username;
     const userInfo = await userModel.getUser(username);
-    userInfo['admin'] = userInfo.grouplist.includes(authConfig.groupConfig.adminGroup.groupname);
-    userInfo['virtualCluster'] = userInfo['extension']['virtualCluster'] ? userInfo['extension']['virtualCluster'] : [];
+    const groupItems = await groupModel.getListGroup(userInfo.grouplist);
+    userInfo['admin'] = groupModel.getAdminWithGroupinfo(groupItems);
+    userInfo['virtualCluster'] = await groupModel.getVCsWithGroupinfo(groupItems);
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/user.js]: getVCListWithGroupInfo will be much better

File: src/rest-server/src/controllers/v2/user.js [hunk 21]
```diff
@@ -40,26 +52,33 @@ const getUser = async (req, res, next) => {
...
   try {
     const userList = await userModel.getAllUser();
-    let retUserList = [];
-    for (let userItem of userList) {
-      userItem['admin'] = userItem.grouplist.includes(authConfig.groupConfig.adminGroup.groupname);
-      userItem['virtualCluster'] = userItem['extension']['virtualCluster'] ? userItem['extension']['virtualCluster'] : [];
-      delete userItem['password'];
-      retUserList.push(userItem);
+    const groupInfo = await groupModel.getAllGroup();
+    const allVClist = Object.keys(await vcModel.prototype.getVcListAsyc());
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/user.js]: vcList or AllVC

File: src/rest-server/src/controllers/v2/user.js [hunk 22]
```diff
@@ -40,26 +52,33 @@ const getUser = async (req, res, next) => {
...
+      delete userItem.password;
+      return userItem;
+    }));
     return res.status(200).json(retUserList);
   } catch (error) {
     return next(createError.unknown(error));
   }
 };
 
+// OIDC
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/user.js]: JSDOC is better

File: src/rest-server/deploy/legacy_user_migrate.py (hunk 10 above)
Comment (mzmssg) [reply src/rest-server/deploy/legacy_user_migrate.py]: They are different funcs, we should make them more genernal. In other words, this func shouldn't assume input is a partly encoded object.

File: src/rest-server/deploy/legacy_user_migrate.py (hunk 12 above)
Comment (mzmssg) [reply src/rest-server/deploy/legacy_user_migrate.py]: No, it will be checked and only be converted once, see line 164

File: src/rest-server/src/controllers/v2/group.js (hunk 13 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/group.js]: I want to provide an api to update single key in a nested extension, like `/extension/acls/virtualClusters` or `/extension/acls/virtualClusters` or further `/extension/attr_a/attr_b/attr_c...`. In express, I could only find the way of match route path by '*' and then reference it by `req.params[0]`. Do you have better suggestion?

File: src/rest-server/src/controllers/v2/group.js (hunk 14 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/group.js]: Because we don't know the updated attrs or its type, here I use data as a general name.

File: src/rest-server/src/controllers/v2/group.js (hunk 15 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/group.js]: Yes, more specificly, only one key in a nested extension

File: src/rest-server/src/controllers/v2/user.js [hunk 23]
```diff
@@ -270,9 +256,6 @@ const addGroupIntoUserGrouplist = async (req, res, next) => {
     const username = req.params.username;
     const groupname = req.body.groupname;
     let userInfo = await userModel.getUser(username);
-    if (userInfo.grouplist.includes(authConfig.groupConfig.adminGroup.groupname)) {
```
Comment (ydye) [inline src/rest-server/src/controllers/v2/user.js]: Allow the modify for admin user?

File: src/rest-server/src/models/v2/group.js (hunk 17 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/group.js]: Yes, getGroupVCs is frequently used, so I provide a wrapper for it.

File: src/rest-server/src/models/v2/group.js (hunk 19 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/group.js]: In currently implementation, there exists a circular dependencies: group model depends on user model (when deleting group), user model depends on group mode (when checking admin and vc). CommonJS can't handle it correctly. As a workaround, we need require model on demand, instead at top.

Comment (mzmssg) [reply src/rest-server/src/models/v2/group.js]: To solve it, I think we could remove the group model dependency, move related logic like deleting user grouplist into controller.

File: src/rest-server/src/config/v2/user.js (hunk 2 above)
Comment (mzmssg) [reply src/rest-server/src/config/v2/user.js]: Create user:
image

Get user:
image

Comment (mzmssg) [reply src/rest-server/src/config/v2/user.js]: For object, we should use `default()` instead of `default({})`, they could achive the same result, using `default()` is less possible to make mistakes.

File: src/rest-server/src/config/v2/group.js (hunk 1 above)
Comment (mzmssg) [reply src/rest-server/src/config/v2/group.js]: Same reason as below

File: src/rest-server/src/utils/manager/user/user.js (hunk 3 above)
Comment (mzmssg) [reply src/rest-server/src/utils/manager/user/user.js]: Same reason as above

File: src/rest-server/src/controllers/v2/group.js (hunk 13 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/group.js]: See https://stackoverflow.com/questions/23876045/get-rest-of-the-path-from-requested-route-in-nodejs

File: src/rest-server/deploy/legacy_user_migrate.py (hunk 11 above)
Comment (mzmssg) [reply src/rest-server/deploy/legacy_user_migrate.py]: See above

File: src/rest-server/src/models/v2/group.js [hunk 24]
```diff
@@ -31,236 +30,182 @@ const crudConfig = crudGroup.initConfig(process.env.K8S_APISERVER_URI);
...
+// TODO: replace updateGroup2ExnternalMapper
+// const updateExternalName2Groupname = async () => {
+//   const groupList = await getAllGroup();
+//   externalName2Groupname.clear();
+//   for (const groupItem of groupList) {
+//     externalName2Groupname[groupItem.externalName] = groupItem.groupname;
+//   }
+// };
+
+// hack for basic mode, assume every vc have a same name group
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: JSDOC

File: src/rest-server/src/models/v2/group.js [hunk 25]
```diff
@@ -358,74 +293,64 @@ const createAdminUser = async () => {
...
+//       logger.info('No user\' grouplist need to be updated.');
+//     }
+//   } catch (error) {
+//     // eslint-disable-next-line no-console
+//     console.log(error);
+//   }
+// };
 
-const groupTypeVCCheck = async () => {
+// delete non-exist vc from all groups
```
Comment (ydye) [inline src/rest-server/src/models/v2/group.js]: jsdoc

File: src/rest-server/src/utils/manager/group/group.js (hunk 4 above)
Comment (mzmssg) [reply src/rest-server/src/utils/manager/group/group.js]: We already have two validation and defaulting: input from route, output to k8s. I'm not sure it's a good idea to introduce more default logic at other level. I think setup all defaulting at the bottom could make upper level more clean
For compatible schema change, I think it's fine, especially we are changing extension field, which is designed for changable.

File: src/rest-server/src/models/v2/group.js (hunk 16 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/group.js]: Workaround for circular dependency, see http://www.ruanyifeng.com/blog/2015/11/circular-dependency.html.

User model require group model, then if we require user model in group model in turn, here will get a empty userModel.

File: src/rest-server/src/controllers/v2/user.js (hunk 23 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/user.js]: There are two questions:
1. Whether allow to update user permission by update grouplist.
Admin is a group attr, not a user attr, so allow to update user admin by updating his groups make sense. If we stick origin experience, I'm fine, but the logic is disallowing to add/remove *admin group* into/from user grouplist.
2. Whether allow to update admin grouplist if no admin role change.
I think it should be. E.g If we support different group storage preference(not permission), we should also allow to customize such preference for admin, the solution is add admin into a specific group configured his preference.

Comment (ydye) [reply src/rest-server/src/controllers/v2/user.js]: sync with @scarlett2018

File: src/rest-server/src/models/v2/group.js (hunk 16 above)
Comment (ydye) [reply src/rest-server/src/models/v2/group.js]: Create issue about this

File: src/rest-server/src/utils/manager/group/group.js (hunk 4 above)
Comment (ydye) [reply src/rest-server/src/utils/manager/group/group.js]: pending design？

File: src/rest-server/src/controllers/v2/user.js (hunk 20 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/user.js]: Renamed getVCsWithGroupInfo

File: src/rest-server/src/controllers/v2/user.js (hunk 21 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/user.js]: Keep allVClist

File: src/rest-server/src/controllers/v2/user.js (hunk 22 above)
Comment (mzmssg) [reply src/rest-server/src/controllers/v2/user.js]: Add jsdoc for specific funcs, the comments will be kept. Will mark all jsdoc comment resolved

File: src/rest-server/src/models/v2/group.js (hunk 16 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/group.js]: After checkin
